src/server/wasm/wasm_registry.cc

```diff
@@ -0,0 +1,135 @@
+// Copyright 2024, DragonflyDB authors.  All rights reserved.
+// See LICENSE for licensing terms.
+//
+
+#include "server/wasm/wasm_registry.h"
+
+#include <server/wasm/api.h>
+
+#include <shared_mutex>
+
+#include "absl/cleanup/cleanup.h"
+#include "absl/strings/str_cat.h"
+#include "base/logging.h"
+#include "io/file_util.h"
+#include "server/wasm/api.h"
+#include "wasm.h"
+#include "wasmtime.h"
+
+namespace dfly::wasm {
+
+WasmRegistry::WasmRegistry() {
+  engine_ = wasm_engine_new();
+  CHECK(engine_);
+  linker_ = wasmtime_linker_new(engine_);
+  CHECK(linker_);
+  CHECK_EQ(wasmtime_linker_define_wasi(linker_), nullptr);
+
+  // There are more configs we should look at. In particular we can spend a little more
+  // time in compiling but generate a more optimized binary as well ass signal handling config.
+  wasi_config_ = wasi_config_new();
+  CHECK(wasi_config_);
+  wasi_config_inherit_argv(wasi_config_);
+  wasi_config_inherit_env(wasi_config_);
+  wasi_config_inherit_stdin(wasi_config_);
+  wasi_config_inherit_stdout(wasi_config_);
+  wasi_config_inherit_stderr(wasi_config_);
+
+  // Define API functions
+  // This should really be done with a separate abstraction. Something line
+  // WasmFunction::Register()
+  std::string mod = "dragonfly";
+  std::string export_f = "hello";
+  wasm_valtype_vec_new_empty(&result_types_);
+  wasm_valtype_vec_new_empty(&arg_types_);
+  hello_handle_ = wasm_functype_new(&arg_types_, &result_types_);
+  auto error =
+      wasmtime_linker_define_func(linker_, mod.data(), mod.size(), export_f.data(), export_f.size(),
+                                  hello_handle_, api::Hello, nullptr, nullptr);
+  CHECK_EQ(error, nullptr);
+}
+
+WasmRegistry::~WasmRegistry() {
+  wasi_config_delete(wasi_config_);
+  wasmtime_linker_delete(linker_);
+  wasm_engine_delete(engine_);
+}
+
+std::string WasmRegistry::Add(std::string_view path) {
+  // 1. Read the wasm file in path
+  auto is_file_read = io::ReadFileToString(path);
+  if (!is_file_read) {
+    return absl::StrCat("File error for path: ", path, " with error ",
+                        is_file_read.error().message());
+  }
+
+  // 2. Setup && compile
+  //  wasmtime_module_t* module = nullptr;
+  //  wasm_byte_vec_t wasm;
+  //  wasm_byte_vec_new_uninitialized(&wasm, is_file_read->size());
+  //  std::memcpy(wasm.data, is_file_read->data(), wasm.size);
+
+  // The following line is problematic, it causes a SEGFAULT when we switch fibers
+  // after the command gets executed
+
+  //  auto error = wasmtime_module_new(engine_, (uint8_t*)wasm.data, wasm.size, &module);
```

Comment:
if we comment this out we crash when we switch fibers @dranikpg

Reply:
how can we switch fibers if there are no preemption points? The mutex below? Let's make everything thread-local or use a spinlock

Reply:
I am not sure, I had also commented out the locks all together. I will play around and ping